### DISCUSSION TITLE
Replacing text with redaction and insert_textbox and fixing reading order

### DISCUSSION THREAD

**Discussion by bkonic** [pymupdf/PyMuPDF#3422] [Q&A] [answered]
Thanks for this awesome library.
I am trying to remove sensitive data from PDFs using redact annotations and then (after _page.apply_redactions(_)) insert replacement text with _page.insert_textbox_ in the same place. 
This works, but the resulting PDF text does not have the intended reading order, e.g. if I replace "Peter Parker" in "Peter Parker is Spiderman" with "Somebody", the resulting reading order (by manual selection in a pdf reader)  would be "is Spiderman Somebody".
What I understand from the documentation and earlier questions this is due to how the manipulations to the PDF are stored. So "Somebody" is not read as part of the original sentence, but as a "new layer".
I can remove all OCR and then redo OCR with an OCR library which fixes the issue (altough this makes things a bit more cumbersome).
**My Question: Is there another (more elegant) way to fix the reading order after inserting text?**
Thanks in advance for any help!

**Comment by JorjMcKie** [comment, marked as answer]
This is a typical Discussions post.

To your question:

There is no way to do it otherwise. Any other solution would require to recreate the PDF. PDF is not Word.
Other elements (text, images, graphics) on the page stay where they are and cannot be changed.
What is more, the emptied space of removed text cannot be resized in any way. New text must fit exactly - and quite some effort is needed to fit replacement text exactly - in terms of font selection, font size, color and insertion point.
Even if all that is done perfectly, the new text is physically **_appended_** to the page's appearance source code (no "new layer") and will therefore be read at last - if using plain / naive text extraction.
Putting extracted text back in reading sequence requires sorting it adequately, based on geometrical information.

**Reply by bkonic** [reply]
Thank you very much for the swift reply, @JorjMcKie . This helps a lot. I will try to create a new PDF from extracting images from the pages (after text replacement)  and doing OCR over them.

**Reply by JorjMcKie** [reply]
Hm, you are willing to pay a high price just for the sake of reading sequence ... 🤷‍♂️

**Reply by bkonic** [reply]
@JorjMcKie Yes, I am painfully aware. But there seem to be no alternatives. My use case is that the resulting PDF can be used by users to copy+paste text or to analyse the PDF with a large language model (like ChatGPT). In both cases the reading order is crucial - especially if I replace sensitive data in the middle of a sentence. The only other way I see is to provide a way for the users to get "text only" where the application tries to find the right reading order and provides the text to the user (or maybe create a text file).

**Reply by JorjMcKie** [reply]
Users copy-pasting stuff from a PDF viewers window is a bit of an unusual use case / luxury.

Otherwise, extracting words `page.get_text("words", sort=True)` and then re-synthesizing the lines by joining the text of words having (approximately) the same bottom coordinate could be a viable way to reproduce reading sequence. That definitely is the way I would follow for LLM requirements.
